symengine/fields.cpp

```diff
@@ -0,0 +1,241 @@
+            }
+        } else {
+            if (p.second == 1) {
+                args.push_back(pow(var_, integer(p.first)));
+            } else {
+                args.push_back(Mul::from_dict(integer(p.second),
+                                              {{var_, integer(p.first)}}));
+            }
+        }
+    }
+    if (poly_.dict_.empty())
+        args.push_back(zero);
+    return args;
+}
+
+void GaloisFieldDict::gf_div(const GaloisFieldDict &o,
+                         const Ptr<GaloisFieldDict> &quo,
+                         const Ptr<GaloisFieldDict> &rem) const
+{
+    SYMENGINE_ASSERT(modulo_ == o.modulo_);
+    if (o.dict_.empty())
+        throw std::runtime_error("ZeroDivisionError");
+    map_uint_mpz dict_divisor = o.dict_;
+    map_uint_mpz dict_out;
+    size_t deg_dividend = dict_.empty() ? 0 : dict_.rbegin()->first;
+    size_t deg_divisor = dict_divisor.rbegin()->first;
+    if (deg_dividend < deg_divisor) {
+        *quo = GaloisFieldDict(dict_out, modulo_);
+        *rem = GaloisFieldDict(dict_, modulo_);
+    } else {
+        dict_out = dict_;
+        integer_class inv;
+        mp_invert(inv, dict_divisor.rbegin()->second, modulo_);
+        integer_class coeff;
+        for (size_t it = deg_dividend; it != size_t(-1); it--) {
+            coeff = dict_out[it];
+            size_t lb = deg_divisor + it > deg_dividend ? 
+                        deg_divisor + it - deg_dividend : 0;
+            size_t ub = std::min(it + 1, deg_divisor);
+            for (size_t j = lb; j < ub; j++) {
+                mp_addmul(coeff, dict_out[it - j + deg_divisor], -dict_divisor[j]);
+            }
+            if (it >= deg_divisor)
+                coeff *= inv;
+            mp_fdiv_r(dict_out[it], coeff, modulo_);
+        }
+        map_uint_mpz dict_rem, dict_quo;
+        for (auto it : dict_out) {
+            if (it.first < deg_divisor)
+                dict_rem[it.first] = it.second;
+            else
+                dict_quo[it.first - deg_divisor] = it.second;
+        }
+        *quo = GaloisFieldDict(dict_quo, modulo_);
+        *rem = GaloisFieldDict(dict_rem, modulo_);
+    }
+}
+
+GaloisFieldDict GaloisFieldDict::gf_lshift(const integer_class n) const
+{
+    map_uint_mpz dict_out;
+    unsigned n_val = mp_get_si(n);
+    for (auto iter : dict_) {
+        dict_out.insert(dict_out.end(), {iter.first + n_val, iter.second});
+    }
+    return GaloisFieldDict(dict_out, modulo_);
+}
+
+void GaloisFieldDict::gf_rshift(const integer_class n,
+                            const Ptr<GaloisFieldDict> &quo,
+                            const Ptr<GaloisFieldDict> &rem) const
+{
+    map_uint_mpz dict_quo, dict_rem;
+    unsigned n_val = mp_get_si(n);
+    for (auto iter : dict_) {
+        if (iter.first >= n_val)
+            dict_quo.insert(dict_quo.end(), {iter.first - n_val, iter.second});
+        else
+            dict_rem.insert(dict_rem.end(), iter);
+    }
+    *quo = GaloisFieldDict(dict_quo, modulo_);
```

Comment:
Constructor would check that `iter.second` is positive which is not necessary right?